feat(wip-0018): include more details and better description (#72)
witnet/WIPs: wip-0018.md

In other words, it creates a string that contains the debug representation of the inner error. The debug representation is created by the Rust compiler, and it may change between Rust versions. And the inner error is supposed to be an implementation detail of witnet-rust, so it doesn't make much sense for it to be part of consensus-critical code. This gives us a good reason to remove this message.

> This message may also appear during the retrieval and aggregation stages. In that case the revealed bytes can be different for different witnesses, so there it is not consensus-critial and changing the message is not a breaking change. Therefore the message argument should only be removed in the tally stage, when serializing the result of the tally script.

Reviewer: ```suggestion
This message may also appear during the retrieval and aggregation stages. In that case, the revealed bytes can be different for different witnesses, so it is not consensus-critical and changing the message is not a breaking change. Therefore the message argument should only be removed in the tally stage when serializing the result of the tally script.
```

### Example

There are two kinds of `UnhandledIntercept` errors, depending on where they are created:

* Created in tally

> 

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

* Created in retrieval/aggregation

#### Created in tally

> This document focuses on those errors that are created in the tally stage, because they are the consensus-critial ones. As an example, see data request `63c1ff0bacd24676bd71fa2bb984e8faee22cd8c73b14b83c57149a792dc315c`. In that request every node reports a different string, and the mode operator returns an error because there is no "most common string". This is the tally using the old validation rules:

Reviewer: ```suggestion
This document focuses on those errors that are created in the tally stage because they are the consensus-critical. As an example, see data request `63c1ff0bacd24676bd71fa2bb984e8faee22cd8c73b14b83c57149a792dc315c`. In that request, every node reports a different string, and the mode operator returns an error because there is no "most common string". This is the tally using the old validation rules:
```

```

#### Created in retrieval/aggregation

> The other kind of `UnhandledIntercept` can be seen in data request `463bb51d32942fb48edb88684a47440d8165bc97a1912acfab220f6de0e87ff5`. In this request a majority of witnesses report an error, and the result of the request is a mode of the different errors, which happens to be an `UnhandledIntercept` error. This is the tally using the old validation rules:

Reviewer: ```suggestion
The other kind of `UnhandledIntercept` can be seen in data request `463bb51d32942fb48edb88684a47440d8165bc97a1912acfab220f6de0e87ff5`. In this request, a majority of witnesses report an error, and the result of the request is a mode of the different errors, which happens to be an `UnhandledIntercept` error. This is the tally using the old validation rules:
```

```
Tally: RadonTypes::RadonError(UnhandledIntercept { inner: None, message: Some("inner: Decode { from: \"cbor::value::Value\", to: \"RadonArray\" }") })
```

And this is the tally using the logic from WIP-0018:

```
Tally: RadonTypes::RadonError(UnhandledInterceptV2 { inner: None })
```

> However, in this case it is impossible to recover the original error, because the error is created in the retrieval stage and we only have the error message, which is removed from the tally result:

Reviewer: ```suggestion
However, in this case, it is impossible to recover the original error because the error is created in the retrieval stage and we only have the error message, which is removed from the tally result:
```

```
Local tally: RadonTypes::RadonError(UnhandledInterceptV2 { inner: None })
```

But in this case the reveals may contain some potentially helpful error message (same as before WIP-0018).

## Specification

Starting from and including the activation epoch:

**(1)**: All the data requests that result in an `UnhandledIntercept` RADON error (error code `0xFF`) at the tally stage MUST serialize this error with no extra arguments.

**(2)**: Any tally that contains an `UnhandledIntercept` RADON error (error code `0xFF`) with arguments will be considered invalid.

> **(3)**: Any reveal that contains an `UnhandledIntercept` RADON error (error code `0xFF` MAY include an optional message argument.

Reviewer 2: Unclosed parenthesis.
